src/libstd/collections/hash/raw/mod.rs

```diff
@@ -0,0 +1,1212 @@
+        RawIter {
+            iter: self.iter.clone(),
+            items: self.items,
+        }
+    }
+}
+
+impl<T> Iterator for RawIter<T> {
+    type Item = Bucket<T>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Bucket<T>> {
+        match self.iter.next() {
+            Some(b) => {
+                self.items -= 1;
+                Some(b)
+            }
+            None => {
+                // We don't check against items == 0 here to allow the
+                // compiler to optimize away the item count entirely if the
+                // iterator length is never queried.
+                debug_assert_eq!(self.items, 0);
+                None
+            }
+        }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.items, Some(self.items))
+    }
+}
+
+impl<T> ExactSizeIterator for RawIter<T> {}
+impl<T> FusedIterator for RawIter<T> {}
+
+/// Iterator which consumes a table and returns elements.
+pub struct RawIntoIter<T> {
+    iter: RawIter<T>,
+    alloc: Option<(NonNull<u8>, Layout)>,
+}
+
+impl<'a, T> RawIntoIter<T> {
+    #[inline]
+    pub fn iter(&self) -> RawIter<T> {
+        self.iter.clone()
+    }
+}
+
+unsafe impl<T> Send for RawIntoIter<T> where T: Send {}
+unsafe impl<T> Sync for RawIntoIter<T> where T: Sync {}
+
+impl<T> Drop for RawIntoIter<T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            // Drop all remaining elements
+            if mem::needs_drop::<T>() {
+                while let Some(item) = self.iter.next() {
+                    item.drop();
+                }
+            }
+
+            // Free the table
+            if let Some((ptr, layout)) = self.alloc {
+                dealloc(ptr.as_ptr(), layout);
+            }
+        }
+    }
+}
+
+impl<T> Iterator for RawIntoIter<T> {
+    type Item = T;
+
+    #[inline]
+    fn next(&mut self) -> Option<T> {
+        unsafe { Some(self.iter.next()?.read()) }
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.iter.size_hint()
+    }
+}
+
+impl<T> ExactSizeIterator for RawIntoIter<T> {}
+impl<T> FusedIterator for RawIntoIter<T> {}
+
+/// Iterator which consumes elements without freeing the table storage.
+pub struct RawDrain<'a, T: 'a> {
+    iter: RawIter<T>,
+
+    // We don't use a &'a RawTable<T> because we want RawDrain to be covariant
+    // over 'a.
+    table: NonNull<RawTable<T>>,
+    _marker: PhantomData<&'a RawTable<T>>,
+}
+
+impl<'a, T> RawDrain<'a, T> {
+    #[inline]
+    pub fn iter(&self) -> RawIter<T> {
+        self.iter.clone()
+    }
+}
+
+unsafe impl<'a, T> Send for RawDrain<'a, T> where T: Send {}
+unsafe impl<'a, T> Sync for RawDrain<'a, T> where T: Sync {}
+
+impl<'a, T> Drop for RawDrain<'a, T> {
+    #[inline]
+    fn drop(&mut self) {
+        unsafe {
+            // Ensure that the table is reset even if one of the drops panic
+            let _guard = guard(self.table, |table| table.as_mut().clear_no_drop());
```

Comment:
Does this even work? I always forget how panicking inside a destructor works.

Regardless, it's a bit disappointing that we have to double-write every control byte in the happy path. One alternative would be to mem::replace(self, Self::new()) in the drain constructor, and swap back here. Then `next` could ignore the control-bytes altogether, and if you mem::forget Drain you end up with an empty collection, having leaked all the contents, similar to Vec (although with an associated memory leak).

The current impl is ~fine though, not something that needs to be fixed to land. Just food for thought.